pkg/api/types.go

```diff
@@ -1192,13 +1195,15 @@ type ServiceSpec struct {
 	// None can be specified for headless services when proxying is not required
 	ClusterIP string `json:"clusterIP,omitempty"`
 
-	// Type determines how the service will be exposed.  Valid options: ClusterIP, NodePort, LoadBalancer
-	Type ServiceType `json:"type,omitempty"`
 
 	// ExternalIPs are used by external load balancers, or can be set by
 	// users to handle external traffic that arrives at a node.
 	ExternalIPs []string `json:"externalIPs,omitempty"`
 
+	// Only applies to Service Type: LoadBalancer
+	// LoadBalancer will get created with the IP specified in this field.
+	LoadBalancerIP string `json:"loadBalancerIP,omitempty"`
```

Comment:
There are a number of facets of this API.

1) service.spec.type = LoadBalancer.  This tells kubernetes "do whatever
you have to do to manifest a load balancer for me from the cloud provider,
assign it a new IP address, and accept traffic on my
service.spec.ports[*].port"

2) service.spec.externalIPs[].  This tells kubernetes "I am managing some
other IPs that will arrive at the nodes in the cluster - please accept
traffic to those IPs on service.spec.ports[*].port and forward it to my
service"

3) service.spec.ports[*].NodePort.  This tells kubernetes "I want to expose
my service on a port on each node of the cluster - please accept that
traffic and forward it to my service"

This PR applies to case (1) above.  If specified, loadBalancerIP is a
request that the cloud provider use the specific IP, rather than a new
one.  Does that make more sense?

From your description, it sounds like a "floating IP" could implement this
API - #12589 #12284

On Sun, Aug 30, 2015 at 7:48 PM, Angus Lees <notifications@github.com>
wrote:

> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/13005#discussion_r38281540>
> :
>
> >
> >  	// ExternalIPs are used by external load balancers, or can be set by
> >  	// users to handle external traffic that arrives at a node.
> >  	ExternalIPs []string `json:"externalIPs,omitempty"`
> >
> > +	// Only applies to Service Type: LoadBalancer
> > +	// LoadBalancer will get created with the IP specified in this field.
> > +	LoadBalancerIP string `json:"loadBalancerIP,omitempty"`
>
> I'm slightly confused by what LoadBalancerIP actually is :/
>
> I thought it was some idea of "chassis IP", which OpenStack doesn't have.
> You *can* refer to loadbalancers by an *ID* however. In some deployments
> these refer to real hardware loadbalancers, so can't be created/destroyed
> by regular users. In LBaaS v2 API, each "loadbalancer" can only have a
> single external IP (aka VIP), so exposing this might not mesh well with
> ExternalIPs[].
>
> Your description above "allocate me an IP that is mine until I release it"
> sounds like you might be referring to something OpenStack calls a "floating
> IP". Yes, you can have a public IP allocated from a pool - and associate
> that with a loadbalancer. And you can pre-allocate the floating-ip before
> configuring any loadbalancer functionality.
>
> I'm still struggling to understand exactly how LoadBalancerIP is intended
> to differ from ExternalIPs. We might indeed be talking about some common
> functionality, but at the moment neither the code comment here, the commit
> description, nor the discussion so far has helped me understand the
> expected semantics beyond "it's an IP that has something to do with the
> loadbalancer" :P
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13005/files#r38281540>.
>